stl/src/vector_algorithms.cpp

```diff
@@ -1327,6 +1399,144 @@ namespace {
+                        constexpr _Ty _Correction = _Ty{1} << (sizeof(_Ty) * 8 - 1);
+
+                        if constexpr ((_Mode & _Mode_min) != 0) {
+                            _Cur_min_val += _Correction;
+                        }
+
+                        if constexpr ((_Mode & _Mode_max) != 0) {
+                            _Cur_max_val += _Correction;
+                        }
+                    }
+
+                    break;
+                }
+                // This is the main part, finding vertical minimum/maximum
+
+                _Cur_vals = _Traits::_Load(_First);
+
+                if constexpr (_Sign_correction) {
+                    _Cur_vals = _Traits::_Sign_correction(_Traits::_Load(_First), false);
+                }
+
+                if constexpr ((_Mode & _Mode_min) != 0) {
+                    if constexpr (_Sign || _Sign_correction) {
+                        _Cur_vals_min = _Traits::_Min(_Cur_vals_min, _Cur_vals); // Update the current minimum
+                    } else {
+                        _Cur_vals_min = _Traits::_Min_u(_Cur_vals_min, _Cur_vals); // Update the current minimum
+                    }
+                }
+
+                if constexpr ((_Mode & _Mode_max) != 0) {
+                    if constexpr (_Sign || _Sign_correction) {
+                        _Cur_vals_max = _Traits::_Max(_Cur_vals_max, _Cur_vals); // Update the current maximum
+                    } else {
+                        _Cur_vals_max = _Traits::_Max_u(_Cur_vals_max, _Cur_vals); // Update the current maximum
+                    }
+                }
+            }
+        } else {
+            _Cur_min_val = *reinterpret_cast<const _Ty*>(_First);
+            _Cur_max_val = *reinterpret_cast<const _Ty*>(_First);
+
+            _Advance_bytes(_First, sizeof(_Ty));
+        }
+
+#endif // !_M_ARM64EC
+        if constexpr (_Mode == _Mode_min) {
+            for (auto _Ptr = static_cast<const _Ty*>(_First); _Ptr != _Last; ++_Ptr) {
+                if (*_Ptr < _Cur_min_val) {
+                    _Cur_min_val = *_Ptr;
+                }
+            }
+            return _Cur_min_val;
+        } else if constexpr (_Mode == _Mode_max) {
+            for (auto _Ptr = static_cast<const _Ty*>(_First); _Ptr != _Last; ++_Ptr) {
+                if (_Cur_max_val < *_Ptr) {
+                    _Cur_max_val = *_Ptr;
+                }
+            }
+            return _Cur_max_val;
+        } else {
+            for (auto _Ptr = static_cast<const _Ty*>(_First); _Ptr != _Last; ++_Ptr) {
+                if (*_Ptr < _Cur_min_val) {
+                    _Cur_min_val = *_Ptr;
+                }
+                // Not else!
+                // * Needed for correctness if start with maximum, as we don't handle specially the first element.
+                // * Promote branchless code generation.
+                if (_Cur_max_val <= *_Ptr) {
+                    _Cur_max_val = *_Ptr;
+                }

```

Comment:
First question: This code was copy-paste-modified, and I've realized I don't understand the "Needed for correctness" part. Can you explain the original scenario?

https://github.com/microsoft/STL/blob/192a84008a59ac4d2e55681e1ffac73535788674/stl/src/vector_algorithms.cpp#L529-L535

Second question: In this new function, that only cares about values instead of locations, the less-than-or-equal test in `if (_Cur_max_val <= *_Ptr)` is unnecessarily overwriting `_Cur_max_val`, right? Shouldn't this be a strictly less-than test now?

Reply:
You are right about the "correctness" part in the new function.

The original `_Minmax_element` function may enter the fallback without any elements processed, so it will have maximum possible value as minimum, and minimum possible value for maximum.

The new `_Minmax` function does one step before entering the fallback, It does that without any extra check, as it takes advantage of the input range not being empty. So it does not apply to the new function.

For the original `_Minmax_element` function, one step before the fallback can be performed too, but it is much less advantageous: it needs empty range check, and also it will not avoid minimum/maximum possible value initialization, so I prefer to keep that as it is.

Reply:
Regarding strict vs non-strict test. You are correct for the high-level code. 

I expect branchless codegen, and I expect it is even more likely with the new `_Minmax` function, so it should not matter in the end, as branchless instructions will always write to the dest register.

Still I'll fix that, I will not try to mimic `_Minmax_element`

Reply:
> The original `_Minmax_element` function may enter the fallback without any elements processed, so it will have maximum possible value as minimum, and minimum possible value for maximum.

Thanks, I forgot about that part!

The new code here is much easier for me to understand, thanks. :heart_eyes_cat: